ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,445 @@
+}
+
+/**
+ * Logs depending on failure type
+ * @param context the Azure functions context
+ * @param failure the failure to log
+ */
+const logFailure = (context: Context) => (
+  failure: ActivityResultFailure
+): void => {
+  switch (failure.kind) {
+    case "INVALID_INPUT_FAILURE":
+      context.log.error(
+        `${logPrefix}|Error decoding input|ERROR=${failure.reason}`
+      );
+      break;
+    case "QUERY_FAILURE":
+      context.log.error(
+        `${logPrefix}|Error ${failure.query} query error |ERROR=${failure.reason}`
+      );
+      break;
+    case "USER_NOT_FOUND_FAILURE":
+      context.log.error(`${logPrefix}|Error user not found |ERROR=`);
+      break;
+    default:
+      assertNever(failure);
+  }
+};
+
+/**
+ * Factory methods that builds an activity function
+ *
+ * @param messageModel
+ * @param notificationModel
+ * @param profileModel
+ * @param senderServiceModel
+ * @param blobService
+ *
+ * @returns an activity function in the form (Context, ActivityInput) -> Promise<Either<ActivityResultFailure, ActivityResultSuccess>>
+ */
+export const createExtractUserDataActivityHandler = (
+  messageModel: MessageModel,
+  notificationModel: NotificationModel,
+  notificationStatusModel: NotificationStatusModel,
+  profileModel: ProfileModel,
+  senderServiceModel: SenderServiceModel,
+  blobService: BlobService
+  // tslint:disable-next-line: no-big-function
+) => {
+  /**
+   * Look for a profile from a given fiscal code
+   * @param fiscalCode a fiscal code identifying the user
+   * @returns either a user profile, a query error or a user-not-found error
+   */
+  const taskifiedFindProfile = (
+    fiscalCode: FiscalCode
+  ): TaskEither<
+    ActivityResultUserNotFound | ActivityResultQueryFailure,
+    Profile
+  > =>
+    fromQueryEither(
+      () => profileModel.findOneProfileByFiscalCode(fiscalCode),
+      "findOneProfileByFiscalCode"
+    ).foldTaskEither<
+      ActivityResultUserNotFound | ActivityResultQueryFailure,
+      Profile
+    >(
+      failure => fromEither(left(failure)),
+      maybeProfile =>
+        fromEither<ActivityResultUserNotFound, Profile>(
+          fromOption(
+            ActivityResultUserNotFound.encode({
+              kind: "USER_NOT_FOUND_FAILURE"
+            })
+          )(maybeProfile)
+        )
+    );
+
+  /**
+   * Retrieves all contents for provided messages
+   * @param messages
+   */
+  const getAllMessageContents = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageContentWithId>
+  > =>
+    array.sequence(taskEither)(
+      messages.map(({ id: messageId }) =>
+        fromQueryEither(
+          () => messageModel.getContentFromBlob(blobService, messageId),
+          "messageModel.getContentFromBlob"
+        ).foldTaskEither<ActivityResultQueryFailure, MessageContentWithId>(
+          failure => fromEither(left(failure)),
```

Comment:
when the left value is unchanged, probably you can `chain` or `map` instead of folding

Reply:
In this case it's not trivial, because the _right_ path may become _left_ if Option is none.